### PR TITLE
FED-2444 Null safety preparation for conditional prop function calls

### PR DESCRIPTION
**Pull request by aaronlademann-wf**
## Motivation
Functions passed in as prop values are often null checked before being called using patterns that do not migrate nicely to null-safety - requiring the use of the `!` assertion like so:

```dart
if (props.onClick != null) {
  props.onClick!(event);
}
// or
if (props.onClick != null) props.onClick!(event);
```

## Changes
1. Add a suggestor that removes the separate conditional null check and instead adds a null-aware `?` operator
1. Add the suggestor to the existing `null_safety_prep` executable.
1. Write tests.

### What it does
**Before**
```dart
if (props.onClick != null) {
  props.onClick(event);
}
// or
if (props.onClick != null) props.onClick(event);
```

**After**
```dart
props.onClick?.call(event);
```

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: test/dart3_suggestors/null_safety_prep/fn_prop_null_aware_call_suggestor_test.dart [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                  }, [props.onClick]);
+                  
+                  return (Dom.button()..onClick = handleClick)();
+                },
+                UiFactoryConfig(displayName: 'Foo'),
+              );
+            '''),
+        );
+      });
+    });
```
Comment (greglittlefield-wf) [inline test/dart3_suggestors/null_safety_prep/fn_prop_null_aware_call_suggestor_test.dart:134]: Could we also add test cases that
- if-checks that don't involve props aren't updated (and the codemod skips over them without throwing)
- then statements that `return` something aren't updated
- then statements that do something besides call the prop aren't updated

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Looking good, left some comments on cases to handle and tests

**Comment by aaronlademann-wf** [comment]
All feedback addressed!

File: lib/src/dart3_suggestors/null_safety_prep/fn_prop_null_aware_call_suggestor.dart [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    final propFunctionBeingNullChecked =
+        _getPropFunctionBeingNullChecked(condition);
+    final ifStatement = parent;
+    if (ifStatement.elseStatement != null) return null;
+    if (ifStatement.parent
+            ?.thisOrAncestorOfType<IfStatement>()
+            ?.elseStatement !=
+        null) {
+      // There is an else-if statement present
+      return null;
```
Comment (greglittlefield-wf) [inline lib/src/dart3_suggestors/null_safety_prep/fn_prop_null_aware_call_suggestor.dart:115]: I think this check isn't as specific as it should be, and due to the `thisOrAncestorOfType`, might cause cases to get skipped if they're nested. For example, I'd expect this to get skipped with the current logic
```dart
if (condition) {
  if (props.onClick != null) props.onClick(e)
} else {}
```

Could we just check the parent directly instead?
```suggestion
    if (ifStatement.parent.tryCast<IfStatement>()?.elseStatement ==
        ifStatement) {
      // ifStatement is an else-if
      return null;
    }
```

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Thanks for addressing that feedback! Just one more small comment, otherwise LGTM; I also tested it locally and it works great!

**Review by greglittlefield-wf — APPROVED** [review]
+10

**Review by greglittlefield-wf — APPROVED** [review]
+10

@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**